allennlp_models/rc/models/bidaf_no_answer.py

```diff
@@ -0,0 +1,386 @@
+from allennlp_models.rc.models.utils import get_best_span
+
+logger = logging.getLogger(__name__)
+
+
+@Model.register("bidaf_no_answer")
+class BidirectionalAttentionFlowNoAnswer(Model):
+    """
+    This class implements Minjoon Seo's `Bidirectional Attention Flow model
+    <https://www.semanticscholar.org/paper/Bidirectional-Attention-Flow-for-Machine-Seo-Kembhavi/7586b7cca1deba124af80609327395e613a20e9d>`_
+    for answering reading comprehension questions (ICLR 2017). In particular, it implements the no-answer
+    variant used in Omer Levy's `Zero-Shot Relation Extraction via Reading Comprehension
+    <https://www.semanticscholar.org/paper/Zero-Shot-Relation-Extraction-via-Reading-Levy-Seo/fa025e5d117929361bcf798437957762eb5bb6d4>`_`
+    paper.
+
+    Parameters
+    ----------
+    vocab : ``Vocabulary``
+    text_field_embedder : ``TextFieldEmbedder``
+        Used to embed the ``question`` and ``passage`` ``TextFields`` we get as input to the model.
+    num_highway_layers : ``int``
+        The number of highway layers to use in between embedding the input and passing it through
+        the phrase layer.
+    phrase_layer : ``Seq2SeqEncoder``
+        The encoder (with its own internal stacking) that we will use in between embedding tokens
+        and doing the bidirectional attention.
+    matrix_attention : ``MatrixAttention``
+        The attention function that we will use when comparing encoded passage and question
+        representations.
+    modeling_layer : ``Seq2SeqEncoder``
+        The encoder (with its own internal stacking) that we will use in between the bidirectional
+        attention and predicting span start and end.
+    span_end_encoder : ``Seq2SeqEncoder``
+        The encoder that we will use to incorporate span start predictions into the passage state
+        before predicting span end.
+    dropout : ``float``, optional (default=0.2)
+        If greater than 0, we will apply dropout with this probability after all encoders (pytorch
+        LSTMs do not apply dropout to their last layer).
+    no_answer_threshold : ``float``, optional (default=None)
+        If the probability of no answer is greater than this value, predict no answer, regardless of whether
+        it is the highest-scoring span or not.
+    initializer : ``InitializerApplicator``, optional (default=``InitializerApplicator()``)
+        Used to initialize the model parameters.
+    regularizer : ``RegularizerApplicator``, optional (default=``None``)
+        If provided, will be used to calculate the regularization penalty during training.
+    """
+
+    def __init__(
+        self,
+        vocab: Vocabulary,
+        text_field_embedder: TextFieldEmbedder,
+        num_highway_layers: int,
+        phrase_layer: Seq2SeqEncoder,
+        matrix_attention: MatrixAttention,
+        modeling_layer: Seq2SeqEncoder,
+        span_end_encoder: Seq2SeqEncoder,
+        dropout: float = 0.2,
+        no_answer_threshold: float = None,
```

Comment:
If this is the same as regular BiDAF, except for this parameter, can we have just one class for both?

Reply:
@nelson-liu I'm happy to make that change if you don't have time, but I'd like to know whether that will work before I do it.

Reply:
Yeah, it's the same, but `forward()` needs to be tweaked significantly...which is why I didn't subclass.